pkg/gitignore/gitignore_test.go

```diff
@@ -48,6 +48,8 @@ func TestGitignoreMatch(t *testing.T) {
 		{Name: "Anywhere in tree, file", Path: "so/many/layers/any.txt", Pattern: "any.txt", Expected: true},
 		{Name: "Anywhere in tree, dir", Path: "so/many/layers/here/", Pattern: "here/", Expected: true},
 		{Name: "Question mark", Path: "a/b/cXd.txt", Pattern: "a/b/c?d.txt", Expected: true},
+		{Name: "Question mark extension A", Path: "a/b/index.ts", Pattern: "a/b/**.tsx?", Expected: true},
```

Comment:
hey @zomars thanks for the pull request!

this gitignore package is supposed to mimic `.gitignore` behavior. based on a couple of documentations i found this test would not match expected behavior `A question mark matches exactly one character.`

- https://www.atlassian.com/git/tutorials/saving-changes/gitignore#:~:text=A%20question%20mark%20matches%20exactly%20one%20character.
- https://git-scm.com/docs/gitignore

let me know if you see otherwise.

Reply:
Thanks for the response @dvdliao that makes sense. Although shouldn't this pattern match the [expected one from Github workflows](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet) so we can just reuse the same paths from actions themselves?

I've been finding several inconsistencies like [this config](https://github.com/zomars/tmp-require-conditional/blob/a45f9697e33e43d457331a91865dd60147103413/.github/workflows/e2e-embed.yml#L6-L8
) should [match this required checks](https://github.com/zomars/tmp-require-conditional/blob/73b89e34a6fe584dc89044a8d0a8f0c94e4b8f7b/.github/workflows/require-conditional.yml#L39-L41) but it doesn't.

Seems like negating a glob children from a previously ignored parent results in adding ALL paths that are not negated files.

So this:

```yaml
paths:
  - /apps/web/**
  - '!/apps/web/public/static/locales/**'
```

is basically equivalent of this:

```yaml
paths:
  - '!/apps/web/public/static/locales/**'
  - /**/* 
```

Am I making sense?

Reply:
Also @dvdliao regarding this

> A question mark matches exactly one character.

Doesn't that means that `tsx?` should match both `ts` and `tsx` ? Or is that a regex thing only?

Reply:
From [GitHub path docs](![image](https://user-images.githubusercontent.com/3504472/177416605-e9bfc8de-3729-4b61-b422-49aa9bea094b.png)):

![image](https://user-images.githubusercontent.com/3504472/177416592-ee80b59c-e3c7-4836-ad86-9c4e08d8b415.png)

Reply:
hey @zomars if i am understanding correctly, i think these series of questions is actually a behavior difference of `gitignore` versus [github filters](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet). 

i do not currently see a behavior issue of the gitignore package. i also did a sanity check in my editor that should cover your case
<img width="519" alt="Screen Shot 2022-07-05 at 3 00 29 PM" src="https://user-images.githubusercontent.com/7452900/177424892-7b0efc89-183c-4df6-8dff-c0cf8f75c30d.png">
and found that it matches the documentation here: https://git-scm.com/docs/gitignore
> It is not possible to re-include a file if a parent directory of that file is excluded. 

at the time of this writing, this action currently follows `gitignore` patterns.

however, i do agree with you that it makes more sense to mimic github filters pattern. i can take it back to the team as a request (#10) but cannot say when/if it will be done. to my understanding it will likely be a breaking change release.

Reply:
Yeah it took me a while to figure it out since I was expecting GH filter patterns and in the README there's no mention of the gitignore convention. 

Please let me know if there's a way for me to help further 🙏